feature/platform/transceiver/tests/zr_firmware_version_test/zr_firmware_version_test.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zr_firmware_version_test
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/openconfig/featureprofiles/internal/deviations"
+	"github.com/openconfig/featureprofiles/internal/fptest"
+	"github.com/openconfig/featureprofiles/internal/samplestream"
+	"github.com/openconfig/ondatra"
+	"github.com/openconfig/ondatra/gnmi"
+	"github.com/openconfig/ondatra/gnmi/oc"
+	"github.com/openconfig/ygot/ygot"
+)
+
+const (
+	dp16QAM           = 1
+	targetOutputPower = -10
+	frequency         = 193100000
+)
+
+func TestMain(m *testing.M) {
+	fptest.RunTests(m)
+}
+
+// Topology: dut:port1 <--> port2:dut
+
+func configInterface(t *testing.T, dut1 *ondatra.DUTDevice, dp *ondatra.Port, enable bool) {
+	d := &oc.Root{}
+	i := d.GetOrCreateInterface(dp.Name())
+	i.Enabled = ygot.Bool(enable)
+	i.Type = oc.IETFInterfaces_InterfaceType_ethernetCsmacd
+	gnmi.Replace(t, dut1, gnmi.OC().Interface(dp.Name()).Config(), i)
+	component := opticalChannelComponentFromPort(t, dut1, dp)
+	gnmi.Replace(t, dut1, gnmi.OC().Component(component).OpticalChannel().Config(), &oc.Component_OpticalChannel{
+		TargetOutputPower: ygot.Float64(targetOutputPower),
+		Frequency:         ygot.Uint64(frequency),
+	})
+}
+func verifyFirmwareVersionValue(t *testing.T, dut1 *ondatra.DUTDevice, pStream *samplestream.SampleStream[string]) {
+	firmwareVersionSample := pStream.Next(t)
+	if firmwareVersionSample == nil {
+		t.Fatalf("Firmware telemetry %v was not streamed in the most recent subscription interval", firmwareVersionSample)
+	}
+	firmwareVersionVal, ok := firmwareVersionSample.Val()
+	if !ok {
+		t.Fatalf("Firmware version %q telemetry is not present", firmwareVersionSample)
+	}
+	// Check firmware version return value of correct type
+	if reflect.TypeOf(firmwareVersionVal).Kind() != reflect.String {
+		t.Fatalf("Return value is not type string")
+	}
+}
+
+func TestZRFirmwareVersionState(t *testing.T) {
+	dut1 := ondatra.DUT(t, "dut")
+	dp1 := dut1.Port(t, "port1")
+	dp2 := dut1.Port(t, "port2")
```

Comment:
(See related testbed comment in metadata.textproto)

Because all the ports on the device are not likely to be 400ZR, should we instead iterate through all the provided ports and find one which is of the desired type (PMD400GBASEZR).  Or are we guaranteed that TESTBED_DUT_400ZR will always return port1 and port2 which are PMD400GBASEZR?

Reply:
Nevermind, I see in our internal reservation system (not part of featureprofiles), the reservation system is expected to provide a device with port1 and port2 already containing PMD400GBASEZR.  I will resolve this comment.